core/src/main/scala/org/apache/spark/ui/scope/RDDOperationGraphListener.scala

```diff
@@ -58,7 +63,16 @@ private[ui] class RDDOperationGraphListener(conf: SparkConf) extends SparkListen
     }
     jobIdToStageIds(jobId) = stageInfos.map(_.stageId).sorted
 
-    // Remove graph metadata for old stages
+    // Remove state for old jobs
+    if (jobIds.size >= retainedJobs) {
```

Comment:
You forget to insert `jobId` to `jobIds`. Could you write a simple test to verify we won't leak anything.